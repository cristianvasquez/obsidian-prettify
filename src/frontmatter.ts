import type {FontmatterInput, MarkdownPrettifierOptions} from "./domain";
import Templates from "./templates";
import {DEFAULT_OPTIONS} from "./constants";
// https://npm.io/package/gray-matter
import matter from 'gray-matter'
import yaml from 'js-yaml'
const templateUtil = new Templates();

function addQuotesToTags(txt) {
    // put quotes around the
    // - #value
    //
    // and tags: #value
    // and tag: #value
    //
    return txt.replace(
        /^(- |tags:|tag:)\s?(#[a-z//\d-_]+)/gi,
        "$1'$2'"
    );
}


async function frontmatter(
    content: string,
    userOptions: MarkdownPrettifierOptions,
    frontmatterInput: FontmatterInput
) {

    let options = {...DEFAULT_OPTIONS, ...userOptions};

    if (content.split('---\n',3).length===3) {
        const frontMatter = content.split('---\n',3).join('---\n')
        const frontMatterQuotes  = addQuotesToTags(frontMatter)
        content.replace(frontMatter,frontMatterQuotes)
    }

    const file = matter(content)

    function applyUUID(data){
        if (frontmatterInput.addUUIDIfNotPresent && !data.id) {
            data.id = templateUtil.getNewUUID()
        }
        return data
    }

    const noData = (file) => Object.keys(file.data).length === 0

    if (noData(file) && options.createHeaderIfNotPresent) {
        file.data = applyValues(options.newHeaderTemplate, frontmatterInput)
    } else if (options.updateHeaders){
        const data = applyValues(options.updateHeaderTemplate, frontmatterInput)
        file.data = mergeValues(file.data, data)
    }

    file.data = applyUUID(file.data)
    return matter.stringify(file.content, file.data).replace(/\n$/, "");
}

function addDashes(template){
    return `---
${template}
---`
}

function applyValues(template: string, frontMatterData: FontmatterInput) {

    const newTags = (opt) => (opt.tags && opt.tags?.length > 0)
    // Update uuids
    template = templateUtil.replaceUUID(template)

    // Update dates and times
    if (frontMatterData.today) {
        template = String(
            templateUtil.replaceDates(template, frontMatterData.today)
        );
    }
    // Add things like new tags
    let doc =  matter(addDashes(template))
    if (newTags(frontMatterData)) {
        doc.data.tags = frontMatterData.tags;
    }
    return doc.data;
}

function getUnion(first: any, second: any): Array<string> {
    let firstArray = first ? first : []
    let secondArray = second ? second : []
    if (typeof firstArray === "string" || firstArray instanceof String) {
        firstArray = [firstArray];
    }
    if (typeof secondArray === "string" || secondArray instanceof String) {
        secondArray = [secondArray];
    }
    return Array.from(
        new Set([...firstArray, ...secondArray]).values()
    );
}

function mergeValues(currentYaml: any, newYAML: any) {

    // Assign all new values
    if (currentYaml) {
        Object.assign(currentYaml, newYAML);
    }

    // Calculate the union of tags
    const tagUnion: Array<string> = getUnion(currentYaml.tags, newYAML.tags)
    if (tagUnion.length > 0) {
        currentYaml.tags = tagUnion;
    }

    // Calculate the union of aliases
    const aliasUnion: Array<string> = getUnion(currentYaml.alias, newYAML.alias)
    if (aliasUnion.length > 0) {
        currentYaml.aliasUnion = aliasUnion;
    }

    return currentYaml;
}


export default frontmatter
