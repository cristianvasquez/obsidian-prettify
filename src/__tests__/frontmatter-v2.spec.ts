import frontmatter from "../frontmatter";
import moment from "moment";
const fixed_date = moment("2010-06-09T15:20:00-07:00");

test("createHeaderIfNotPresent: false", () => {
  const content = `No Header`;
  return frontmatter(
    content,
    { createHeaderIfNotPresent: false },
    { today: fixed_date }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("createHeaderIfNotPresent: true", () => {
  const content = `No Header`;
  return frontmatter(
    content,
    {
      createHeaderIfNotPresent: true,
    },
    { today: fixed_date }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("createHeaderIfNotPresent: true", () => {
  const content = `No Header`;
  const newHeaderTemplate =`
id: {{UUID}}
`

  return frontmatter(
      content,
      {
        createHeaderIfNotPresent: true,
        newHeaderTemplate: newHeaderTemplate
      },
      { today: fixed_date }
  ).then((data) => {
    expect(data).not.toContain('id: {{UUID}}')
  });
});

test("Do not update if it's a new header", () => {
  const content = `No Header`;
  return frontmatter(
      content,
      {
        createHeaderIfNotPresent: true,
        updateHeader:true
      },
      { today: fixed_date }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("Creates date first time", () => {
  const content = `No Header`;
  return frontmatter(
    content,
    {
      createHeaderIfNotPresent: true,
      updateHeader: false,
    },
    { today: fixed_date }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("Does not create header if update but not create", () => {
  const content = `No Header`;
  return frontmatter(
      content,
      {
        createHeaderIfNotPresent: false,
        updateHeader: true,
      },
      { today: fixed_date }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});


test("Does not modify date", () => {
  const content = `---
date updated: '1999-06-10T00:20:00+02:00'

---`;
  return frontmatter(
    content,
    {
      updateHeader: false,
    },
    { today: fixed_date }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("Maintains properties in header", () => {
  const content = `---
title: 'Example'
list:
  - oneX
  - 0
  - false
  - 'quotes'
---
# A header

some text
            `;
  return frontmatter(content, {}, { today: fixed_date }).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("Respects hash", () => {
  const content = `---
tag: '#Interpretability'

---
            `;
  return frontmatter(content, {}, { today: fixed_date }).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("Respects emojis", () => {
  const content = `---
tag: '#🦐'

---🦐
            `;
  return frontmatter(content, {}, { today: fixed_date }).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

test("add quote", () => {
  const content = `---
tag: #hello

---
            `;
  return frontmatter(
    content,
    {},
    {
    }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

it("Add quotes", () => {
  const content = `---
Name: 
 - #Hello
 - World
 - ' #Hi'
 - #with/tab
 - #with/tab/teb/tib
 - #howdy_bla
 - #howdy-bla
Other tag: #Nothing
tags: #Nothing
---

Outside frontmatter

#tag

`;
  return frontmatter(content, {}, { today: fixed_date }).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

it("Adds tags", () => {
  const content = `---
Name: 
 - #Hi
---


`;
  return frontmatter(
    content,
    {},
    { today: fixed_date, tags: ["#hello", "#world"] }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

it("Preserve previous tags", () => {
  const content = `---
tags: 
 - #preserve_me
---

`;
  return frontmatter(
    content,
    {},
    { today: fixed_date, tags: ["#hello", "#world"] }
  ).then((data) => {
    expect(data).toMatchSnapshot();
  });
});

it("Preserve previous tags", () => {
  const content = `---
date updated: '2020-12-09T00:14:06+01:00'

tags: '#literatura'

---
`;
  return frontmatter(content, {}, { today: fixed_date, tags: [] }).then(
    (data) => {
      expect(data).toMatchSnapshot();
    }
  );
});



it("Preserve previous tags", () => {
const content = `---
date: 2021-02-05--5pmt5:20:48:42

---`;
  return frontmatter(content, {}, { today: fixed_date, tags: [] }).then(
    (data) => {
      expect(data).toMatchSnapshot();
    }
  );
});

// UUID


it("Adds a new UUID, existing frontmatter", () => {
  const content = `---
date: 2021-02-05--5pmt5:20:48:42
---`;
  return frontmatter(content, {}, {
    today: fixed_date,
    tags: [],
    addUUIDIfNotPresent: true
  }).then(
      (data) => {
        expect(data).toMatchSnapshot();
      }
  );
});

it("Adds a new UUID", () => {
  const content = `
  No FrontMatter 
  `;
  return frontmatter(content, {
    createHeaderIfNotPresent: true
  }, {
    today: fixed_date,
    tags: [],
    addUUIDIfNotPresent: true
  }).then(
      (data) => {
        expect(data).toMatchSnapshot();
      }
  );
});

it("Adds a new UUID, newHeaderTemplate", () => {
  const content = `
  No FrontMatter  
  `;
  return frontmatter(content, {
    createHeaderIfNotPresent:true,
    newHeaderTemplate: ''
  }, {
    tags: [],
    addUUIDIfNotPresent: true
  }).then(
      (data) => {
        expect(data).toMatchSnapshot();
      }
  );
});


it("maintain UUID", () => {
  const content = `---
date: 2021-02-05--5pmt5:20:48:42
id: alreadyPresent
---`;
  return frontmatter(content, {}, {
    today: fixed_date,
    tags: [],
    addUUIDIfNotPresent: false
  }).then(
      (data) => {
        expect(data).toMatchSnapshot();
      }
  );
});

it("Multiple fences", () => {
  const content = `---
date: 2021-02-05--5pmt5:20:48:42
id: alreadyPresent
---
---
other: something else
---
`;
  return frontmatter(content, {}, {
    today: fixed_date,
    tags: [],
    addUUIDIfNotPresent: false
  }).then(
      (data) => {
        expect(data).toMatchSnapshot();
      }
  );
});


it("No newHeaderTemplate", () => {
  const content = `
  No FrontMatter  
  `;
  return frontmatter(content, {
    createHeaderIfNotPresent:false,
    newHeaderTemplate: ''
  }, {
    tags: [],
    addUUIDIfNotPresent: true
  }).then(
      (data) => {
        expect(data).toMatchSnapshot();
      }
  );
});
