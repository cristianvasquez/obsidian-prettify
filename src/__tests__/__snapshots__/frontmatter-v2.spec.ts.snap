// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add quotes 1`] = `
"---
Name: 
 - #Hello
 - World
 - ' #Hi'
 - #with/tab
 - #with/tab/teb/tib
 - #howdy_bla
 - #howdy-bla
Other tag: #Nothing
tags: #Nothing
---

Outside frontmatter

#tag

"
`;

exports[`Adds a new UUID 1`] = `
"---
date: 2021-02-05--5pmt5:20:48:42
---"
`;

exports[`Adds a new UUID 2`] = `
"---
date created: '2010-06-10 00:20'
id: c5627cfc-3bb6-4185-93cf-6b3cc7c10fa8
---

"
`;

exports[`Adds a new UUID, existing frontmatter 1`] = `
"---
date: 2021-02-05--5pmt5:20:48:42
---"
`;

exports[`Adds a new UUID, newHeaderTemplate 1`] = `
"---
id: c031791b-cf9c-463a-9d4a-84ed9584cb68
---

  No FrontMatter  
  
"
`;

exports[`Adds tags 1`] = `
"---
Name: 
 - #Hi
---


"
`;

exports[`Creates date first time 1`] = `
"---
date created: '2010-06-10 00:20'
---
No Header"
`;

exports[`Do not update if it's a new header 1`] = `
"---
date created: '2010-06-10 00:20'
---
No Header"
`;

exports[`Does not create header if update but not create 1`] = `"No Header"`;

exports[`Does not modify date 1`] = `
"---
date updated: '1999-06-10T00:20:00+02:00'
---
"
`;

exports[`Maintains properties in header 1`] = `
"---
title: Example
list:
  - oneX
  - 0
  - false
  - quotes
---
# A header

some text
            "
`;

exports[`Multiple fences 1`] = `
"---
other: something else
date: '2021-02-05--5pmt5:20:48:42'
id: alreadyPresent
---
"
`;

exports[`No newHeaderTemplate 1`] = `
"---
id: 72214964-4c9e-437b-bf4c-bb4ee0feb3ba
---

  No FrontMatter  
  "
`;

exports[`Preserve previous tags 1`] = `
"---
tags:
  - null
---
"
`;

exports[`Preserve previous tags 2`] = `
"---
date updated: '2020-12-09T00:14:06+01:00'
tags: '#literatura'
---
"
`;

exports[`Preserve previous tags 3`] = `
"---
date: '2021-02-05--5pmt5:20:48:42'
---
"
`;

exports[`Respects emojis 1`] = `
"---
tag: '#ü¶ê'

---ü¶ê
            "
`;

exports[`Respects hash 1`] = `
"---
tag: '#Interpretability'
---
            "
`;

exports[`add quote 1`] = `
"---
tag: #hello

---
            "
`;

exports[`createHeaderIfNotPresent: false 1`] = `"No Header"`;

exports[`createHeaderIfNotPresent: true 1`] = `
"---
date created: '2010-06-10 00:20'
---
No Header"
`;

exports[`maintain UUID 1`] = `
"---
date: 2021-02-05--5pmt5:20:48:42
id: alreadyPresent
---"
`;
